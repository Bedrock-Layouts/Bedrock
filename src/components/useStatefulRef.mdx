---
name: useStatefulRef
menu: Components
---

# useStatefulRef

React RefObjects are not stateful, or in other words, changing the `current` property on the RefObject doesn't trigger a rerender.
One can use a ref callback to updated state, but now we are no longer using RefObjects and there is great power in consistency and knowing we will be getting a
RefObject every time. `useStatefulRef` will return a RefObject that can be passed around and used just like any other RefObject, except that changes to the `current`
property will trigger a rerender just like updating state.

## Use case

Anytime you want a RefObject that will trigger a rerender when the `current` property is updated

## How to install

```bash
  npm install @bedrock-layout/use-stateful-ref
```

## Basic usage

```javascript
const App = () => {
  const ref = useStatefulRef(null);
  //... Use the ref object will trigger a rerender
  return <Div ref={ref}>{...content}</Div>;
};
```

## Storybook

[Checkout storybook](https://storybook.bedrock-layout.dev/?path=/story/usestatefulref--basic) for more details.
