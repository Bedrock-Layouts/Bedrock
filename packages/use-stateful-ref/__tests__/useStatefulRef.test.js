import React from "react";
import ReactDOM from "react-dom/";
import { act } from "react-dom/test-utils";

import useStatefulRef from "../src";

jest.spyOn(React, "useState");

let state = undefined;

const setState = jest.fn((newState) => {
  state = newState;
});

React.useState.mockReturnValue([state, setState]);

let statefulRef;
const HookWrapper = ({ value, initialValue }) => {
  statefulRef = useStatefulRef(initialValue);
  if (value) {
    statefulRef.current = value;
  }
  return null;
};

describe("useStatefulRef", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    statefulRef = undefined;
    document.body.removeChild(container);
    container = null;
    setState.mockRestore();
  });

  test("useStatefulRef is not null", () => {
    expect(useStatefulRef).toBeTruthy();
  });

  it("should call useState", () => {
    act(() => {
      ReactDOM.render(<HookWrapper />, container);
    });

    expect(React.useState).toBeCalled();
  });

  it("should call setState when setting value", () => {
    act(() => {
      ReactDOM.render(
        <HookWrapper initialValue="Testing" value="Test" />,
        container
      );
    });

    expect(setState).toBeCalled();
    expect(statefulRef).toMatchSnapshot();
  });

  it("should call not call setState when setting value to same value", () => {
    act(() => {
      ReactDOM.render(
        <HookWrapper initialValue="Test" value="Test" />,
        container
      );
    });

    expect(setState.mock.calls).toMatchSnapshot();
  });
});
