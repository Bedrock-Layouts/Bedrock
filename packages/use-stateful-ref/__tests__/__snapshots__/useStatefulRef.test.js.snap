// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStatefulRef should call not call setState when setting value to same value 1`] = `
Array [
  Array [
    "Test",
  ],
]
`;

exports[`useStatefulRef should call setState when setting value 1`] = `
Object {
  "current": "Test",
}
`;
